pkg/client/restclient/request.go

```diff
@@ -728,14 +733,15 @@ func (r *Request) Stream() (io.ReadCloser, error) {
 	r.backoffMgr.Sleep(r.backoffMgr.CalculateBackoff(r.URL()))
 	resp, err := client.Do(req)
 	updateURLMetrics(r, resp, err)
-	if r.baseURL != nil {
+	if r.urlProvider.Get() != nil {
```

Comment:
I don't like calling r.urlProvider.Get() as a way to know whether r.URL() would return nil.  I would prefer this to be the behavior:

```
if url := r.URL(); url == nil {
   if err != nil {
     r.backoffMgr.UpdateBackoff(url, ...)
     ...
```

It's not obvious to a casual browser that r.urlProvider.Get() and r.URL() are coupled - one should dominate / hide the details of the other.